hal/src/samd21/adc.rs

```diff
@@ -26,18 +26,37 @@ impl Adc<ADC> {
         while adc.status.read().syncbusy().bit_is_set() {}
         adc.sampctrl.modify(|_, w| unsafe { w.samplen().bits(5) }); //sample length
         while adc.status.read().syncbusy().bit_is_set() {}
-        adc.inputctrl.modify(|_, w| w.muxneg().gnd()); // No negative input (internal gnd) 
+        adc.inputctrl.modify(|_, w| w.muxneg().gnd()); // No negative input (internal gnd)
         while adc.status.read().syncbusy().bit_is_set() {}
-        adc.avgctrl.modify(|_, w| {
-            w.samplenum()._1(); // No averaging
-            unsafe { w.adjres().bits(0) } // adjust result by 0
+
+        let mut newadc = Self { adc };
+        newadc.samples(adc::avgctrl::SAMPLENUM_A::_1);
+        newadc.gain(adc::inputctrl::GAIN_A::DIV2);
+        newadc.reference(adc::refctrl::REFSEL_A::INTVCC1);
+
+        newadc
+    }
+
+    pub fn samples(&mut self, samples: adc::avgctrl::SAMPLENUM_A) {
+        self.adc.avgctrl.modify(|_, w| {
+            w.samplenum().variant(samples);
+            // I don't see any reason to divide ourselves. peripheral will automatically
+            // shift as needed
+            unsafe { w.adjres().bits(0) }
```

Comment:
@sajattack are you talking about this comment? I was just saying I feel we don't need to expose an API to adjust `adjres` since that does additional divides if you oversample, but the peripheral automatically divides as needed if the resulting resolution is >16 bits. I left your gain adjustment as is in the default case